File: starterkit2/python/TestControl.py
# coding=utf-8

class TestControl :

    # public
    def Start(self) :
        self.Goto(self.S_START)
        return
    
    def IsEnd(self) :
        return self.CheckState(self.S_END)

    # Manager
    m_curfunc  = None
    m_nextfunc = None

    m_noWait = False

    def Update(self) :
        while(True) :
            bFirst = False
            self.m_noWait = False
            if self.m_nextfunc != None :
                self.m_curfunc = self.m_nextfunc
                self.m_nextfunc = None
                bFirst = True
                #print("start of " + self.m_curfunc.__name__)

            if self.m_curfunc != None :
                self.m_curfunc(bFirst)
            
            if self.m_noWait == False :
                return

    def Goto(self,st) :
        self.m_nextfunc = st
        return

    def CheckState(self, st) :
        return self.m_curfunc == st

    def HasNextState(self) :
        return self.m_nextfunc != None

    def NoWait(self) :
        self.m_noWait = True

    #gosub
    MAX_CALL_STACK = 10
    #              1    2    3    4    5    6    7    8    9    10
    m_callstack = [None,None,None,None,None,None,None,None,None,None]
    m_callstack_level = 0 
    def GoSubState(self,sub,nex) :
        if self.m_callstack_level >= self.MAX_CALL_STACK :
            print("CALL STACK OVER FLOW")
            exit()
        self.m_callstack[self.m_callstack_level] = nex
        self.m_callstack_level = self.m_callstack_level + 1
        self.Goto(sub)

    def ReturnState(self) :
        if self.m_callstack_level <= 0 :
            print("CALL STACK UNDER FLOW")
            exit()
        self.m_callstack_level = self.m_callstack_level - 1
        st = self.m_callstack[self.m_callstack_level]
        self.Goto(st) 

    # [SYN-G-GEN OUTPUT START] indent(4) $/./$
    #             psggConverterLib.dll converted from psgg-file:TestControl.psgg

    #    S_0000
    #
    def S_0000(self,bFirst) :
        if (bFirst) :
            print("Hello World.")
        if self.HasNextState()==False :
            self.Goto(self.S_RET000)
        return
    #    S_0001
    #
    def S_0001(self,bFirst) :
        if (bFirst) :
            print("Loop count." + str(self.m_i) )
        if self.HasNextState()==False :
            self.Goto(self.S_RET001)
        return
    #    S_END
    #
    def S_END(self, bFirst) :
        # nothing to do
        return
    #    S_GSB000
    #
    def S_GSB000(self, bFirst) :
        self.GoSubState(self.S_SBS000,self.S_LOP000)
        self.NoWait()
        return
    #    S_LOP000
    #
    m_i = 0
    def S_LOP000(self, bFirst) :
        self.m_i = 0
        self.Goto(self.S_LOP000_LoopCheckAndGosub____)
        self.NoWait()
        return
    def S_LOP000_LoopCheckAndGosub____(self, bFirst) :
        if  self.m_i < 10 :
            self.GoSubState(self.S_SBS001,self.S_LOP000_LoopNext____)
        else :
            self.Goto(self.S_END)
        self.NoWait()
        return
    def S_LOP000_LoopNext____(self, bFirst) :
        self.m_i = self.m_i + 1
        self.Goto(self.S_LOP000_LoopCheckAndGosub____)
        self.NoWait()
        return
    #    S_RET000
    #
    def S_RET000(self, bFirst) :
        self.ReturnState()
        self.NoWait()
        return
    #    S_RET001
    #
    def S_RET001(self, bFirst) :
        self.ReturnState()
        self.NoWait()
        return
    #    S_SBS000
    #
    def S_SBS000(self,bFirst) :
        if self.HasNextState()==False :
            self.Goto(self.S_0000)
        return
    #    S_SBS001
    #
    def S_SBS001(self,bFirst) :
        if self.HasNextState()==False :
            self.Goto(self.S_0001)
        return
    #    S_START
    #
    def S_START(self, bFirst) :
        self.Goto(self.S_GSB000)
        self.NoWait()
        return


    # [SYN-G-GEN OUTPUT END]

    #m_bYesNo = False
    #def br_YES(self,st) :
    #    if self.HasNextState()==False :
    #        if self.m_bYesNo==True :
    #            self.Goto(st)
    #    return

    #def br_NO(self,st) :
    #    if self.HasNextState()==False :
    #        if self.m_bYesNo==False :
    #            self.Goto(st)

File: starterkit2/php/TestControl.php
<?php

class TestControl {

    // :::: PSGG MANAGER ::::

    private $m_curfunc  = NULL;
    private $m_nextfunc = NULL;
    private $m_noWait   = FALSE;

    private const MAX_CALL_STACK = 10;
    private $m_callstack = array();
    private $m_callstack_level = 0;

    function __construct() {
        $m_callstack = array_fill(0,self::MAX_CALL_STACK,'');
    }

    public function Update() {
        while(TRUE) {
            $bFirst = FALSE;
            if ($this->m_nextfunc!=NULL) {
                $this->m_curfunc = $this->m_nextfunc;
                $this->m_nextfunc = NULL;
                $bFirst = TRUE;
            }
            $this->m_noWait = FALSE;
            if ($this->m_curfunc!=NULL) {
                $this->call_state($this->m_curfunc, $bFirst);
            }
            if ($this->m_noWait == FALSE) {
                break;
            }
        }
    }

    public function GotoState($st) {
        $this->m_nextfunc = $st;
    }

    public function CheckState($st) {
        return $st == $this->m_curfunc;
    }
 
    public function HasNextState() {
        return $this->m_nextfunc != NULL;
    }

    public function NoWait() {
        $this->m_noWait = TRUE;
    }

    public function GoSubState($sub, $next) {
        if ($this->m_callstack_level >= self::MAX_CALL_STACK - 1) {
            echo 'CALL STACK OVER FLOW<br>';
            exit();
        }
        $this->m_callstack[$this->m_callstack_level++] = $next;
        $this->GotoState($sub);
    }

    public function ReturnState()
    {
        if ($this->m_callstack_level <= 0) {
            echo 'CALL STACK UNDER FLOW<br>';
            exit();
        }
        $this->m_callstack_level--;
        $st = $this->m_callstack[$this->m_callstack_level];
        
        $this->GotoState($st);        
    }
    
    function Start() {
        $this->GotoState('S_START');
    }

    function IsEnd() {
        return $this->CheckState('S_END');
    }

    // :::: PSG OUPUT ::::

    function call_state($state, $bFirst) {
        switch($state) {
            //[PSGG OUTPUT START] indent(12) $/^S_/->#case$
            //             psggConverterLib.dll converted from psgg-file:TestControl.psgg

            case 'S_0008': $this->S_0008($bFirst); break;
            case 'S_END': $this->S_END($bFirst); break;
            case 'S_GSB000': $this->S_GSB000($bFirst); break;
            case 'S_GSB001': $this->S_GSB001($bFirst); break;
            case 'S_LOP000': $this->S_LOP000($bFirst); break;
            case 'S_LOP000_LoopCheckAndGosub____': $this->S_LOP000_LoopCheckAndGosub____($bFirst); break;
            case 'S_LOP000_LoopNext____': $this->S_LOP000_LoopNext____($bFirst); break;
            case 'S_RET000': $this->S_RET000($bFirst); break;
            case 'S_SBS000': $this->S_SBS000($bFirst); break;
            case 'S_START': $this->S_START($bFirst); break;


            //[PSGG OUTPUT END]
        }
    }

    //[PSGG OUTPUT START] indent(4) $/./$
    //             psggConverterLib.dll converted from psgg-file:TestControl.psgg

    /*
        S_0008
    */
    function S_0008($bFirst) {
        if ($bFirst)
        {
            echo 'Hello Subroutine.<br>';
        }
        if ($this->HasNextState()==FALSE)
        {
            $this->GotoState('S_RET000');
        }
    }
    /*
        S_END
    */
    function S_END($bFirst)
    {
    }
    /*
        S_GSB000
    */
    function S_GSB000($bFirst)
    {
        $this->GoSubState('S_SBS000','S_GSB001');
        $this->NoWait();
    }
    /*
        S_GSB001
    */
    function S_GSB001($bFirst)
    {
        $this->GoSubState('S_SBS000','S_END');
        $this->NoWait();
    }
    /*
        S_LOP000
    */
    private $index = 0;
    function S_LOP000($bFirst)
    {
        $this->index = 0;
        $this->GotoState('S_LOP000_LoopCheckAndGosub____');
        $this->NoWait();
    }
    function S_LOP000_LoopCheckAndGosub____($bFirst)
    {
        if ($this->index < 10) $this->GoSubState('S_SBS000','S_LOP000_LoopNext____');
        else               $this->GotoState('S_GSB000');
        $this->NoWait();
    }
    function S_LOP000_LoopNext____($bFirst)
    {
        $this->index ++;
        $this->GotoState('S_LOP000_LoopCheckAndGosub____');
        $this->NoWait();
    }
    /*
        S_RET000
    */
    function S_RET000($bFirst)
    {
        $this->ReturnState();
        $this->NoWait();
    }
    /*
        S_SBS000
    */
    function S_SBS000($bFirst) {
        if ($this->HasNextState()==FALSE)
        {
            $this->GotoState('S_0008');
        }
    }
    /*
        S_START
    */
    function S_START($bFirst)
    {
        $this->GotoState('S_LOP000');
        $this->NoWait();
    }


    //[PSGG OUTPUT END]

    // :::: WRITE YOUR CODE HERE ::::

    protected $m_bYesNo;

    function br_YES($st) {
        if ($this->HasNextState()==FALSE) {
            if ($this->m_bYesNo==TRUE) {
                $this->GotoState($st);
            }
        }
    }

    function br_NO($st) {
        if ($this->HasNextState()==FALSE) {
            if ($this->m_bYesNo==FALSE) {
                $this->GotoState($st);
            }
        }
    }
}

/* :::: PSGG MACRO ::::
:psgg-macro-start

; This section has macro defines for converting.

; commentline format  {%0} will be replaced to a comment.
commentline=// {%0} 

@branch=@@@
<<<?"{%0}"/^brifc{0,1}$/
if ([[brcond:{%N}]]) { $this->GotoState( '{%1}' ); }
>>>
<<<?"{%0}"/^brelseifc{0,1}$/
elseif ([[brcond:{%N}]]) { $this->GotoState( '{%1}' ); }
>>>
<<<?"{%0}"/^brelse$/
else { $this->GotoState( '{%1}' ); }
>>>
<<<?"{%0}"/^br_/
$this->{%0}('{%1}');
>>>
@@@

#case=@@@
case '[[state]]': $this->[[state]]($bFirst); break;
<<<?state-typ/^loop$/
case '[[state]]_LoopCheckAndGosub____': $this->[[state]]_LoopCheckAndGosub____($bFirst); break;
case '[[state]]_LoopNext____': $this->[[state]]_LoopNext____($bFirst); break;
>>>
@@@

:psgg-macro-end
*/

?>